### PR TITLE
[SYSDM] Show real hardware name in the System properties

### PR DESCRIPTION
**Pull request by binarymaster**
## Purpose

Show real hardware name in the System properties. :relaxed: 

JIRA issue: None

## Proposed changes

- Initialize output buffer with empty string if DMI string is NULL
- Move system information controls 16 points higher
- Add one more IDC_MACHINELINE
- Implement nice printing of hardware model / motherboard name

## Screenshots

**VirtualBox:**

image

**Real hardware:**

image image

### CODE REVIEW

**Comment by HBelusca** [comment]
I'm wondering whether it wouldn't be better to use a single EDITTEXT zone in read-only mode with flat borderless style to print the strings (separated by newlines); this would also fix the cut build string that we can see on the screenshots above.

**Comment by binarymaster** [comment]
@HBelusca nice idea! It would also allow to copy the text into clipboard. But I think I'll do it later in a separate PR.

**Comment by learn-more** [comment]
Filling the empty space also sounds like something for a new PR.

**Comment by binarymaster** [comment]
Alright, let it be so :)

File: dll/cpl/sysdm/general.c [hunk 1]
```diff
@@ -417,14 +419,330 @@ static VOID SetProcSpeed(HWND hwnd, HKEY hKey, LPTSTR Value, UINT uID)
...
+        HeapFree(GetProcessHeap(), 0, ls2);
+    }
+    return result;
+}
+
+/*
+ * Replaces full word with another shorter word
+ */
+static
+VOID wcsrep(
```
Comment (learn-more) [inline dll/cpl/sysdm/general.c]: It's a shame this is not a c++ file.

File: dll/cpl/sysdm/general.c [hunk 2]
```diff
@@ -417,14 +419,330 @@ static VOID SetProcSpeed(HWND hwnd, HKEY hKey, LPTSTR Value, UINT uID)
...
+    GetSMBiosStringW(DmiStrings[SYS_VENDOR], ven, _countof(ven), TRUE);
+    GetSMBiosStringW(DmiStrings[SYS_PRODUCT], dev, _countof(dev), TRUE);
+
+    if (wcslen(dev) == 0 ||
+        !wcscmp(dev, ven) ||
+        !wcscmp(dev, L"All Series") || // some ASUS boards
+        !wcscmp(ven, L"To be filled by O.E.M.") || // various boards
+        !wcscmp(dev, L"To be filled by O.E.M.") || // various boards
+        !wcscmp(ven, L"To Be Filled By O.E.M.") || // various boards
+        !wcscmp(dev, L"To Be Filled By O.E.M.") || // various boards
```
Comment (learn-more) [inline dll/cpl/sysdm/general.c]: use wcsicmp?

Comment (learn-more) [reply dll/cpl/sysdm/general.c]: Also, not sure how many more are going to be here, might think about adding them in a struct/ array/ whatever

File: dll/cpl/sysdm/general.c [hunk 3]
```diff
@@ -417,14 +419,330 @@ static VOID SetProcSpeed(HWND hwnd, HKEY hKey, LPTSTR Value, UINT uID)
...
+
+        if (wcslen(ven) == 0 && wcslen(dev) == 0)
+        {
+            // board strings are empty, use BIOS vendor string
+            GetSMBiosStringW(DmiStrings[BIOS_VENDOR], ven, _countof(ven), TRUE);
+        }
+    }
+
+    // make vendor strings shorter
+    wcsrep(ven, L"ASUSTeK", L"ASUS");
```
Comment (learn-more) [inline dll/cpl/sysdm/general.c]: Does windows do this?

Comment (binarymaster) [reply dll/cpl/sysdm/general.c]: I don't know whether Windows does this at all, clearly a new feature :)

File: dll/cpl/sysdm/general.c (hunk 1 above)
Comment (binarymaster) [reply dll/cpl/sysdm/general.c]: Yeah indeed, so I have to implement all that functions.

**Comment by binarymaster** [comment]
@Extravert-ir told me about his doubts that Windows does not make vendor strings shorter (displaying them as-is), and System properties (`sysdm.cpl`) is not the right place for such experiments.

However I think it would be really nice improvement towards the user experience and few other points:
- Most users are expecting to see simple and short string describing the hardware or motherboard model (instead of long and not so simply understandable or unrelated words)
- More difference between ReactOS and Windows (since we're talking about design, not API)
- `dxdiag` and `systeminfo` modules are already fully mimics the Windows' behaviour

Currently I tested this code with 251 unique SMBIOS dumps (mainly from real hardware, but some from virtual machines) collected with help from the community.

**Comment by learn-more** [comment]
The downside of translating those strings is that it becomes harder to diagnose a problem.
The first step if we get a hardware description from someone, becomes finding out where they got it from,
because if it comes from this page it will be edited by us.

**Comment by binarymaster** [comment]
> The downside of translating those strings is that it becomes harder to diagnose a problem.

@learn-more  I don't think we lose much data, check out my comparison table: https://docs.google.com/spreadsheets/d/17psq53hoEtZmgdtdnkBNjOXe6KNXYVwRdWiWdSSQfgg/edit?usp=sharing

> The first step if we get a hardware description from someone, becomes finding out where they got it from, because if it comes from this page it will be edited by us.

If someone would request support and attach a screenshot of this dialog, almost in any situation we have to ask additional question about the hardware.

And as you can see, with my word processing code, in most cases it displays even more precise model name. :wink:

File: dll/cpl/sysdm/general.c [hunk 4]
```diff
@@ -417,14 +431,393 @@ static VOID SetProcSpeed(HWND hwnd, HKEY hKey, LPTSTR Value, UINT uID)
...
+    {
+        StringCchCopyW(ven, _countof(ven), L"MSI");
+    }
+
+    // trim redundant characters
+    TrimCommas(ven);
+    TrimCommas(dev);
+
+    if (!memcmp(ven, dev, wcslen(ven) * sizeof(WCHAR)) ||
+        wcsistr(dev, ven) == dev ||
```
Comment (binarymaster) [inline dll/cpl/sysdm/general.c]: `memcmp` check is not needed here, `wcsistr` does the job.

File: dll/cpl/sysdm/general.c [hunk 5]
```diff
@@ -417,14 +431,393 @@ static VOID SetProcSpeed(HWND hwnd, HKEY hKey, LPTSTR Value, UINT uID)
...
+    // make vendor strings shorter
+    for (i = 0; i < _countof(LongNames); i++)
+    {
+        wcsrep(ven, LongNames[i].pwLongName, LongNames[i].pwShortName);
+    }
+
+    // remove redundant words
+    for (i = 0; i < _countof(RedundantWords); i++)
+    {
+        wcsrep(ven, RedundantWords[i], L"");
```
Comment (learn-more) [inline dll/cpl/sysdm/general.c]: What if the word is part of another word?
Like, `Inc` -> `Incorporated` -> `corporated`

Comment (binarymaster) [reply dll/cpl/sysdm/general.c]: `wcsrep` replaces full word by adding a space at begin and end of string.

File: dll/cpl/sysdm/smbios.c [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+
+    if (wcslen(dev) == 0 ||
+        !wcscmp(dev, ven) ||
+        bGenericName)
+    {
+        // system strings are unusable, use board strings
+        if (DmiStrings[BOARD_VENDOR] != NULL || !bGenericName)
+        {
+            GetSMBiosStringW(DmiStrings[BOARD_VENDOR], ven, _countof(ven), TRUE);
+            GetSMBiosStringW(DmiStrings[BOARD_NAME], dev, _countof(dev), TRUE);
```
Comment (binarymaster) [inline dll/cpl/sysdm/smbios.c]: Check here for generic name too.

File: dll/cpl/sysdm/smbios.c [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+    if (wcsistr(dev, ven) == dev ||
+        (!wcscmp(ven, L"HP") && wcsstr(dev, L" by HP ") != NULL))
+    {
+        // device string contains vendor string, use second only
+        StringCchCopyW(pBuf, cchBuf, dev);
+    }
+    else
+    {
+        StringCchCopyW(pBuf, cchBuf, ven);
+        if (wcslen(pBuf) > 0)
```
Comment (binarymaster) [inline dll/cpl/sysdm/smbios.c]: If `dev` string is empty, we don't need space separator.

File: dll/cpl/sysdm/smbios.c [hunk 8]
```diff
@@ -0,0 +1,593 @@
...
+
+    HeapFree(GetProcessHeap(), 0, pwsReplace);
+freeFind:
+    HeapFree(GetProcessHeap(), 0, pwsFind);
+freeStr:
+    HeapFree(GetProcessHeap(), 0, pwsStr);
+}
+
+static
+BOOL isGenericSystemName(PCWSTR ven, PCWSTR dev)
```
Comment (HBelusca) [inline dll/cpl/sysdm/smbios.c]: "*I*sGenericSystemName" ?

**Pull request merged**